examples/source_separation/conv_tasnet/train.py

```diff
@@ -0,0 +1,175 @@
+                "the sample rate from the check point ({checkpoint['sample_rate']}).")
+        if args.num_speakers != checkpoint['num_speakers']:
+            raise ValueError(
+                "The provided #of speakers ({args.num_speakers}) does not match "
+                "the #of speakers from the check point ({checkpoint['num_speakers']}.)"
+            )
+        start_epoch = checkpoint['epoch']
+
+    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+    _LG.info("Using: %s", device)
+
+    model = conv_tasnet.model.ConvTasNet(
+        num_speakers=args.num_speakers, enc_kernel_size=args.sample_rate * 2 // 1000
+    )
+    model.to(device)
+
+    model = torch.nn.parallel.DistributedDataParallel(model, device_ids=[device])
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.learning_rate)
+
+    if args.resume:
+        _LG.info('Loading parameters from the checkpoint...')
+        model.module.load_state_dict(checkpoint['model'])
+        optimizer.load_state_dict(checkpoint['optimizer'])
+    else:
+        dist_utils.synchronize_params(str(args.save_dir / f"tmp.pt"), model, optimizer)
+
+    _LG.info_on_master("Model:\n%s", model)
+
+    lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(
+        optimizer, factor=0.5, patience=3
+    )
+
+    train_loss_func = conv_tasnet.metrics.PITLoss(
+        loss_func=conv_tasnet.metrics.neg_si_snr
+    )
+
+    train_dataset, eval_dataset = dataset_utils.get_dataset(
+        args.dataset, args.dataset_dir, args.num_speakers
+    )
+    collate_fn = dataset_utils.get_collate_fn(args.dataset)
+
+    train_loader = torch.utils.data.DataLoader(
+        train_dataset,
+        batch_size=args.batch_size,
+        sampler=torch.utils.data.distributed.DistributedSampler(train_dataset),
+        collate_fn=collate_fn,
+    )
+    eval_loader = torch.utils.data.DataLoader(
+        eval_dataset,
+        batch_size=args.batch_size,
+        sampler=torch.utils.data.distributed.DistributedSampler(eval_dataset),
+        collate_fn=collate_fn,
+    )
+
+    trainer = conv_tasnet.trainer.Trainer(
+        model,
+        optimizer,
+        train_loader,
+        eval_loader,
+        train_loss_func,
+        args.grad_clip,
+        device,
+    )
+
+    _LG.info_on_master("Running %s epochs", args.epochs)
+    for epoch in range(start_epoch, start_epoch + args.epochs):
+        _LG.info_on_master("-" * 70)
+        _LG.info_on_master("Epoch: %s", epoch)
+        _LG.info_on_master("Learning rate: %s", optimizer.param_groups[0]["lr"])
+        _LG.info_on_master("-" * 70)
+
+        trainer.train_one_epoch()
+        loss = trainer.evaluate()
+        lr_scheduler.step(loss)
+
+        save_path = args.save_dir / f"epoch_{epoch}.pt"
+        dist_utils.save_on_master(
+            {
+                "model": model.module.state_dict(),
+                "optimizer": optimizer.state_dict(),
+                "num_speakers": args.num_speakers,
```

Comment:
I've found it useful to also serialize the `args` together with the model. It allows for simpler `resume`ing of the model, as we can load all the default args etc

Reply:
I think saving `args` as a record can be a good idea, but making the training script governed by the dumped value makes the training script less flexible. The dumped file could outlive the original script. What if I want to run training with different parameters? What if I want to change the signatures? One good thing about this script not in library but in example directory is that this script does not have to consider backward compatibility (as long as the training script works on all the commits). So making resume behavior depend on `args` in file goes against it.